### PR TITLE
Split build bundle

### PR DESCRIPTION
**Pull request by lubej**
## Description

- use exact version of `ethers` 6.10.0 due to `@oasisprotocol/sapphire-paratime` dependency
- Add `rollup-plugin-visualizer`, use `npx vite-bundle-visualizer` inside `frontend` folder, for report

## Resources

Previously:

```console
❯ npx vite-bundle-visualizer
vite v5.1.4 building for production...
✓ 1238 modules transformed.
dist/index.html                                                      0.47 kB │ gzip:   0.30 kB
dist/assets/figtree-latin-ext-wght-normal-Bptc-cl5.woff2            10.10 kB
dist/assets/roboto-mono-vietnamese-wght-normal-B7nCECYT.woff2       10.26 kB
dist/assets/roboto-mono-greek-wght-normal-C58MjUu-.woff2            14.09 kB
dist/assets/roboto-mono-cyrillic-wght-normal-CK9U7V71.woff2         18.60 kB
dist/assets/figtree-latin-wght-normal-tyJnQ6lj.woff2                20.08 kB
dist/assets/roboto-mono-latin-ext-wght-normal-BWJwKVm9.woff2        22.87 kB
dist/assets/roboto-mono-latin-wght-normal-DGRqvGGI.woff2            32.94 kB
dist/assets/roboto-mono-cyrillic-ext-wght-normal-DPu6aEAn.woff2     35.89 kB
dist/assets/index-f5ltzcoX.css                                      18.48 kB │ gzip:   4.33 kB
dist/assets/index-CKw_9xB2.js                                    1,240.81 kB │ gzip: 410.95 kB

(!) Some chunks are larger than 500 kB after minification. Consider:
- Using dynamic import() to code-split the application
- Use build.rollupOptions.output.manualChunks to improve chunking: https://rollupjs.org/configuration-options/#output-manualchunks
- Adjust chunk size limit for this warning via build.chunkSizeWarningLimit.
✓ built in 2.95s
```

Screenshot_27-3-2024_14617_

Current(all assets under default 500kb limit):

```console
❯ npx vite-bundle-visualizer
vite v5.1.4 building for production...
✓ 1123 modules transformed.
dist/index.html                                                    0.86 kB │ gzip:   0.39 kB
dist/assets/figtree-latin-ext-wght-normal-Bptc-cl5.woff2          10.10 kB
dist/assets/roboto-mono-vietnamese-wght-normal-B7nCECYT.woff2     10.26 kB
dist/assets/roboto-mono-greek-wght-normal-C58MjUu-.woff2          14.09 kB
dist/assets/roboto-mono-cyrillic-wght-normal-CK9U7V71.woff2       18.60 kB
dist/assets/figtree-latin-wght-normal-tyJnQ6lj.woff2              20.08 kB
dist/assets/roboto-mono-latin-ext-wght-normal-BWJwKVm9.woff2      22.87 kB
dist/assets/roboto-mono-latin-wght-normal-DGRqvGGI.woff2          32.94 kB
dist/assets/roboto-mono-cyrillic-ext-wght-normal-DPu6aEAn.woff2   35.89 kB
dist/assets/index-BzmJrU6M.css                                    18.52 kB │ gzip:   4.33 kB
dist/assets/lodash-ByOxUBs_.js                                    28.11 kB │ gzip:  11.56 kB │ map:   228.29 kB
dist/assets/index-awQyAFrF.js                                     47.58 kB │ gzip:  16.91 kB │ map:   129.14 kB
dist/assets/react-dom-Ddi3USIR.js                                167.92 kB │ gzip:  52.75 kB │ map:   504.61 kB
dist/assets/ethers-DN1VZNda.js                                   175.44 kB │ gzip:  51.72 kB │ map:   801.14 kB
dist/assets/recharts-DGNO4Vbe.js                                 216.23 kB │ gzip:  51.07 kB │ map:   857.16 kB
dist/assets/vendor-C7c7JVNO.js                                   386.25 kB │ gzip: 147.39 kB │ map: 1,674.81 kB
✓ built in 2.92s
```
Screenshot_27-3-2024_14413_

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -22,14 +23,15 @@
...
+    "ethers": "6.10.0",
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
     "react-intersection-observer": "^9.8.1",
     "react-parallax-tilt": "^1.7.217",
     "react-responsive": "^9.0.2",
     "react-router-dom": "^6.22.2",
-    "recharts": "^2.12.2"
+    "recharts": "^2.12.2",
+    "rollup-plugin-visualizer": "^5.12.0"
```
Comment (lukaw3d) [inline frontend/package.json:34]: devDependencies

File: frontend/vite.config.ts [hunk 2]
```diff
@@ -1,8 +1,30 @@
...
+          if (id.includes('recharts')) {
+            return 'recharts'
+          }
+          if (id.includes('lodash')) {
+            return 'lodash'
+          }
+          if (id.includes('react-dom')) {
+            return 'react-dom'
+          }
+        },
```
Comment (lukaw3d) [inline frontend/vite.config.ts:26]: What is the benefit of this? It doesn't look like it will load these modules on demand

File: frontend/package.json [hunk 3]
```diff
@@ -22,14 +23,15 @@
     "@oasisprotocol/dapp-voting-backend": "workspace:^",
     "@oasisprotocol/sapphire-paratime": "^1.3.2",
     "@phosphor-icons/core": "^2.0.8",
-    "ethers": "^6.11.1",
+    "ethers": "6.10.0",
```
Comment (lukaw3d) [inline frontend/package.json:26]: Nice! I would separate this into its own commit and note which dependency already loads this version of ethers (this deduplicates ethers dependency)

File: frontend/vite.config.ts (hunk 2 above)
Comment (lubej) [reply frontend/vite.config.ts:26]: That was not my intention. Vite build process complains(just warning though, see first output in PR desc) if a single bundled .js file, is over 500kb uncompressed. It is just a way to specify splitting https://rollupjs.org/configuration-options/#output-manualchunks

But I can look into to dynamically load `recharts`, as it is only used in results page.

File: frontend/package.json (hunk 3 above)
Comment (lubej) [reply frontend/package.json:26]: Kinda mentioned this is due to `@oasisprotocol/sapphire-paratime` dependency in the PR desc. But sure, I can split up the commit.

File: frontend/vite.config.ts (hunk 2 above)
Comment (lubej) [reply frontend/vite.config.ts:26]: Did create dynamic import for `recharts`(https://github.com/oasisprotocol/dapp-votee/tree/ml/split-chunks-dynamic-recharts), but then tree shaking doesn't apply:
Screenshot_28-3-2024_61135_

Also did try to dynamically import pages (https://github.com/oasisprotocol/dapp-votee/tree/ml/split-chunks-dynamic-pages) but that creates too much chunks IMO:
Screenshot_28-3-2024_6152_

I would just go with the current solution.

Comment (lukaw3d) [reply frontend/vite.config.ts:26]: So far I'm convinced that the current solution doesn't do anything beneficial:
- loads and parses the same amount of javascript
- it adds a little overhead with multiple connections
- it adds a little complexity to the configuration
- it prevents vite from analyzing and showing a useful-but-ignorable warning to consider dynamic loading

Am I missing something and are you seeing faster load time?

Comment (lubej) [reply frontend/vite.config.ts:26]: I can try tomorrow and will post the results, but essentially, I am guessing that load time should be a bit faster(still very small app) - due to parallel fetching of the resources.

Comment (lubej) [reply frontend/vite.config.ts:26]: Comparison:

`master`:
image

`ml/split-chunks`:
image

There is like +-100ms difference. Which could as well be within margin of an "error". I would need to perform tests multiple times to see the actual difference.

Comment (lukaw3d) [reply frontend/vite.config.ts:26]: On deployed site I measure:
- https://dapp-votee-try-vite-default-6054235.pages.dev/
  unthrottled network tab Finish: 480ms
  Lighthouse Largest Contentful Paint 1.6s - 2.4s
- https://dapp-votee-try-vite-manual-split-6054235.pages.dev/
  unthrottled network tab Finish: 477ms
  Lighthouse Largest Contentful Paint 1.4s - 2.6s :shrug:

Comment (lubej) [reply frontend/vite.config.ts:26]: image

To me it's quite similar... So merge as is? Or do I revert to single chunk for .js? Because it looks like, it doesn't matter anyway.

Comment (lukaw3d) [reply frontend/vite.config.ts:26]: I would still revert: "chunk should not be larger than 500kB" in commit message is misleading

Comment (lubej) [reply frontend/vite.config.ts:26]: Done

**Review by lukaw3d — APPROVED** [review]
